<a href="https://colab.research.google.com/github/mptorpo/analitica_negocios/blob/main/Reto1_MiguelPimientoRestrepo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


**RETO 1: Modelo Naive Bayes**

0. Se procede con la carga de las librerias de trabajo

In [ ]:
import numpy as np #Funciones numéricas transformación y analisis de datos
import pandas as pd #Conectar el sistema con las de datos

1. Se procede con la carga y organización de la base de datos

In [ ]:
nxl='/content/1. BD2_CarCrash_Int.xlsx' #Ruta de acceso BD
XDB=pd.read_excel(nxl,sheet_name=0) #DataFrame: Tabla de Datos

XDB = XDB.dropna() #Eliminando de plano los registros que poseen datos faltantes
XDB = XDB[XDB["County"] == "SAN FRANCISCO"] #Concentración en datos: condado de San Francisco
XDB["CrashType"] = pd.Categorical(XDB["CrashType"]).codes #Convertir a Valor Numérico

#Se cargan las variables de interes
XDB = XDB[['County','Weekday','ViolCat','ClearWeather','Month','Daylight','CrashType','Highway']]

XD = XDB[['Weekday','ViolCat','ClearWeather','Month','CrashType','Daylight']] #Aquí estan las variables de entrada
yd = XDB['Highway'] #Aquí esta la variable de referencia

County  Weekday  ViolCat  ClearWeather  Month  Daylight  \
907     SAN FRANCISCO        7        1             1      3         1   
1580    SAN FRANCISCO        2        9             1      2         0   
1581    SAN FRANCISCO        7        3             1      1         0   
1586    SAN FRANCISCO        7        3             1      3         0   
1587    SAN FRANCISCO        3        3             1      3         0   
...               ...      ...      ...           ...    ...       ...   
112598  SAN FRANCISCO        5       11             1      8         0   
112599  SAN FRANCISCO        5       11             1      3         1   
112600  SAN FRANCISCO        6       10             0      4         0   
112601  SAN FRANCISCO        4       10             1      4         0   
112602  SAN FRANCISCO        5       12             1      2         1   

        CrashType  Highway  
907             0        1  
1580            0        0  
1581            0        0  
1586            0        0  
1587            0        0  
...           ...      ...  
112598          6        0  
112599          6        0  
112600          6        0  
112601          6        0  
112602          6        0  

[2081 rows x 8 columns]

2. Implementamos el Modelo

In [ ]:
from sklearn.naive_bayes import GaussianNB #Cargar GaussianNB

mnb=GaussianNB() #Aplicar Naive Bayes
mnb.fit(XD,yd) #Aplicar Naive Bayes teniendo en cuenta variables

#Se obtiene la info del modelo
ct=mnb.class_count_
print("Como están separados los datos:", ct)
print("El total de los datos es:", len(XD))
print("Los datos porcentualmente estan div:", ct/len(XD))
print(" ") #Espacio (para estética)

#Se obtiene la info de las variables (media y desviación est)
u=mnb.theta_
print("Las medias de las varibles:\n", u)
print(" ") #Espacio (para estética)
std=mnb.var_ #Esto es varianza
std=np.sqrt(std) #Esto es desviación estandar (sqrt es raiz)
print("La desviación estandar de las variables\n:", std)
print(" ") #Espacio (para estética)

#Se identifican los valores superiores e inferiores de los ingresos
print("EL limite superior de los valores:\n", u+std)
print(" ") #Espacio (para estética)
print("El limite inferior de los valores:\n", u-std)
print(" ") #Espacio (para estética)

Como están separados los datos: [1580.  501.]
El total de los datos es: 2081
Los datos porcentualmente estan div: [0.75925036 0.24074964]
 
Las medias de las varibles:
 [[3.88417722 7.45696203 0.86265823 4.35189873 3.21772152 0.66265823]
 [3.9261477  4.1497006  0.79840319 6.69461078 2.27744511 0.62075848]]
 
La desviación estandar de las variables
: [[1.95062542 3.52136997 0.34420782 2.3433645  1.87371962 0.47280262]
 [2.01603723 2.16198902 0.40119266 3.44163641 1.05765448 0.48519831]]
 
EL limite superior de los valores:
 [[ 5.83480263 10.978332    1.20686605  6.69526323  5.09144114  1.13546084]
 [ 5.94218494  6.31168962  1.19959585 10.13624719  3.33509959  1.10595679]]
 
El limite inferior de los valores:
 [[1.9335518  3.93559205 0.51845041 2.00853424 1.3440019  0.18985561]
 [1.91011047 1.98771158 0.39721054 3.25297437 1.21979063 0.13556017]]
 


**Matriz de Confusión**
3. Se procede con la evaluación (desempeño/exactitud del modelo)

In [ ]:
ydp=mnb.predict(XD) #Pronostico de los positivos y negativos

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)
print(" ") #Espacio (para estética)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
Ex=(VN+VP)/(VN+FP+FN+VP)
print("La exactitud del modelo es:", Ex) #Exactitud/Desempeño General: (Verdadreo Negativo + Verdadero Positivo) / Total Datos | (+ usual)

TE=(FN+FP)/(VN+FP+FN+VP) #Tasa de Error
print("La tasa de error es:", TE)

Sen=(VP/(VP+FN)) #Sensibilidad: Comportamiento del modelo (1)
print("La sensibilidad es:", Sen)

Esp=(VN/(VN+FP)) #Especificidad: Comportamiento del modelo (0) | (+ usual)
print("La especificidad es:", Esp)

pre=(VP/(VP+FP)) #Precisión: Comportamiento general del modelo identificando los + (1) | (+ usual)
print("La presición es:", pre)

pren=(VN/(VN+FN)) #Precisión Negativa: Comportamiento general del modelo identificando los - (0)
print("La presición negativa es:", pren)

[[1365  215]
 [ 224  277]]
 
La exactitud del modelo es: 0.7890437289764536
La tasa de error es: 0.21095627102354636
La sensibilidad es: 0.5528942115768463
La especificidad es: 0.8639240506329114
La presición es: 0.5630081300813008
La presición negativa es: 0.8590308370044053


4. Categoría a la que pertenece una persona con características asignadas

In [ ]:
XS=[[5,1,1,3,4,0]]
mnb.predict(XS)
print("La predicción del modelo es:", mnb.predict(XS))

La predicción del modelo es: [0]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


**Preguntas (Car Crash)**

 1.⁠ ⁠¿Cuál es el número de datos total de la base de datos?
- Después de hacer la limpieza de los registros que poseen datos faltantes, quedamos con un total de **2081** filas y 8 columnas.

-

 2.⁠ ⁠¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?
- 1580 (76%) de los accidentes automovilísticos no ocurrieron en la autopista (Highway), y 501 (24%) sí.

-

 3.⁠ ⁠Determinar los intervalos en los cual se encuentra la variación del día de la semana para cada una de las categorías de clasificación:
- El límite superior de la variación del día de la semana para 0 es 5.83. El límite inferior de la variación del día de la semana para 0 es 1.93. El límite superior de la variación del día de la semana para 1 es 5.94. El límite inferior de la variación del día de la semana para 1 es 1.91.

-

 4.⁠ ⁠Determinar los intervalos en los cual se encuentra la variación de la ViolCat para cada una de las categorías de clasificación:
- El límite superior de la variación del día de la ViolCat para 0 es 10.98. El límite inferior de la variación del día de la ViolCat para 0 es 3.94. El límite superior de la variación del día de la ViolCat para 1 es 6.31. El límite inferior de la variación del día de la ViolCat para 1 es 1.99.

-

 5.⁠ ⁠Determinar el intervalo en los cuales se encuentra la variación de la Categoría de accidentes (CrashType) para cada una de las categorías de clasificación:
- El límite superior de la variación del día de la semana para 0 es 5.09. El límite inferior de la variación del día de la semana para 0 es 1.34. El límite superior de la variación del día de la semana para 1 es 3.34. El límite inferior de la variación del día de la semana para 1 es 1.22.

-

 6.⁠ ⁠¿Cuál es la exactitud alcanzada por el modelo?
- La exactitud alcanzada por el modelo es de 79%

-

 7.⁠ ⁠¿Cuál es la especificidad alcanzada por el modelo?
- La especificidad alcanzada por el modelo es de 86%

-

 8.⁠ ⁠Indicar el intervalo en donde se ubica la cantidad de verdaderos negativos:
- Cantidad de verdaderos negativos: 1365

-

 9.⁠ ⁠Indicar el intervalo en donde se ubica la cantidad de verdaderos positivos:
- Cantidad de verdaderos positivos: 277

-

10.⁠ ⁠¿Cuál es la tasa de error alcanzada por el modelo?
- La tasa de erros alcanzada por el modelo es de 21%.

-

11.⁠ ⁠¿Cuál es la sensibilidad alcanzada por el modelo?
- La sensibilidad alcanzada por el modelo es de 55%

-

12.⁠ ⁠¿Cuál es la precisión alcanzada por el modelo?
- La precisión alcanzada por el modelo es de 56%

-

13.⁠ ⁠¿Indicar la categoría a la que pertenece una persona con las siguientes características [5,1,1,3,4,0]?
- Una persona con las siguientes características [5,1,1,3,4,0] pertenece a la categoría 0, lo cual significa que tendría menos probabilidades de tener un accidente automovilístico en la autopista (Highway).

-

**Análisis General:** El estudio de la información de accidentes de tráfico en el condado de San Francisco muestra patrones claves que afectan la determinación de si un accidente sucede en una carretera o no. Se tomaron en cuenta factores como el día laboral, la intensidad del impacto, las condiciones meteorológicas, el mes, la categoría del accidente y la luz solar durante el día. Después de la purificación de los datos y la transformación de variables categóricas, se aplicó un modelo de Naive Bayes. Este demostró una precisión del 79% (superior a 75%), lo que señala un rendimiento óptimo en la categorización de los accidentes. Además, se determinó que el modelo posee una especificidad del 86%, lo que implica que posee una gran capacidad para identificar adecuadamente los accidentes que no tuvieron lugar en autopistas. No obstante, su sensibilidad del 55% indica que todavía existe un espacio para mejorar en la identificación de casos positivos.